src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -473,6 +475,56 @@ static MemorySegment allocateNative(long bytesSize, long alignmentBytes) {
         return NativeMemorySegmentImpl.makeNativeSegment(bytesSize, alignmentBytes);
     }
 
+    /**
+     * Returns a new native memory segment with given base address and size; the returned segment has its own temporal
+     * bounds, and can therefore be closed; closing such a segment can optionally result in calling an user-provided cleanup
+     * action. This method can be very useful when interacting with custom native memory sources (e.g. custom allocators,
+     * GPU memory, etc.), where an address to some underlying memory region is typically obtained from native code
+     * (often as a plain {@code long} value).
+     * <p>Since the cleanup action is an {@link AutoCloseable} object, another memory segment can be used to specify a
+     * cleanup operation; this can be useful in cases where the user would like to take an existing native segment
+     * and tweak some of the properties of the original segment - as demonstrated below:
+     * <blockquote><pre>{@code
+    MemorySegment segment = MemorySegment.allocateNative(100);
+    MemorySegment unconfinedSegment = MemorySegment.ofNativeUnsafe(
+                                              segment.baseAddress(),
+                                              segment.byteSize(),
+                                              null,
+                                              segment,
+                                              null);
```

Comment:
In such use case, close() called on either segment will affect the other as well, correct?
I am assuming in use case like this, the new segment close is preferred to the old segment and developer most likely should not keep reference to the old segment?

Reply:
Yes - calling close will also affect the other segment. We have three options here:
* forget about passing a segment - just accept a Runnable; but this will make it harder e.g. to take an existing mapped segment and make it "unconfined" it. Also, still doesn't remove the problem, since you can have a native memory segment for an address X, and you can create another unchecked memory segment at same address which calls Unsafe::freeMemory when you close - which will again invalidate the segment
* make it so that when you pass a segment as an argument, the input segment is also killed - so that it becomes "not alive"; this can be bypassed in the same way described above
* leave it as is

There's no great solution here; I think I like either the current approach, or the first one, where we just take a `Runnable` and leave it up to the user to define what should happen when you close.